File: .claude/hooks/pre-write-main-guard.mjs
#!/usr/bin/env node
// @hookstack pre-write-main-guard
// PreToolUse Write|Edit: bloque la première écriture sur main si aucun worktree n'est actif
import { readFileSync } from 'fs';
import { execSync } from 'child_process';
import { fileURLToPath } from 'url';

function defaultExec(cmd) {
  try { return execSync(cmd, { encoding: 'utf8', timeout: 5_000 }).trim(); } catch { return ''; }
}

export function run(input, { exec = defaultExec } = {}) {
  const branch = exec('git branch --show-current') || exec('git rev-parse --abbrev-ref HEAD');
  if (!branch || !/^(main|master)$/.test(branch)) return null;

  const worktreeList = exec('git worktree list');
  const currentRoot = exec('git rev-parse --show-toplevel');
  const mainRoot = worktreeList.split('\n')[0]?.split(/\s+/)[0] ?? '';
  if (mainRoot !== currentRoot) return null;

  const filePath = input.tool_input?.file_path ?? '(fichier inconnu)';

  // Autoriser les écritures vers des fichiers hors du repo principal
  if (filePath !== '(fichier inconnu)' && !filePath.startsWith(`${mainRoot}/`)) return null;

  // Autoriser les écritures dans un worktree secondaire (ex: .claude/worktrees/session-xxx/…)
  const secondaryWorktrees = worktreeList.split('\n')
    .slice(1)
    .map(line => line.split(/\s+/)[0])
    .filter(Boolean);
  if (secondaryWorktrees.some(wt => filePath.startsWith(`${wt}/`))) return null;

  return {
    decision: 'block',
    reason: `Écriture sur \`${branch}\` bloquée : vous êtes sur la branche principale.\nCréez un worktree (\`git worktree add ../mon-fix -b feat/mon-fix\`) ou changez de branche avant de modifier \`${filePath}\`.`,
  };
}

/* v8 ignore next 5 */
if (process.argv[1] === fileURLToPath(import.meta.url)) {
  const input = JSON.parse(readFileSync(0, 'utf8'));
  const result = run(input);
  if (result) process.stdout.write(JSON.stringify(result));
}

File: .claude/hooks/pre-write-main-guard.py
#!/usr/bin/env python3
# @hookstack pre-write-main-guard
"""Blocks the first write on main if no worktree is active (PreToolUse Write|Edit)."""
import json
import re
import subprocess
import sys


def _exec(cmd):
    try:
        return subprocess.run(
            cmd, shell=True, capture_output=True, text=True, timeout=5, check=True
        ).stdout.strip()
    except Exception:
        return ""


def run(input_data, *, exec_cmd=None):
    if exec_cmd is None:
        exec_cmd = _exec

    branch = exec_cmd("git branch --show-current") or exec_cmd(
        "git rev-parse --abbrev-ref HEAD"
    )
    if not branch or not re.match(r"^(main|master)$", branch):
        return None

    worktree_list = exec_cmd("git worktree list")
    current_root = exec_cmd("git rev-parse --show-toplevel")
    lines = worktree_list.split("\n")
    main_root = lines[0].split()[0] if lines and lines[0].strip() else ""
    if main_root != current_root:
        return None

    file_path = (input_data.get("tool_input") or {}).get("file_path") or "(fichier inconnu)"

    # Autoriser les écritures vers des fichiers hors du repo principal
    if file_path != "(fichier inconnu)" and not file_path.startswith(f"{main_root}/"):
        return None

    # Autoriser les écritures dans un worktree secondaire (ex: .claude/worktrees/session-xxx/…)
    secondary_worktrees = [
        line.split()[0]
        for line in lines[1:]
        if line.strip()
    ]
    if any(file_path.startswith(f"{wt}/") for wt in secondary_worktrees):
        return None

    return {
        "decision": "block",
        "reason": (
            f"Écriture sur `{branch}` bloquée : vous êtes sur la branche principale.\n"
            f"Créez un worktree (`git worktree add ../mon-fix -b feat/mon-fix`) ou changez de branche avant de modifier `{file_path}`."
        ),
    }


if __name__ == "__main__":
    data = json.load(sys.stdin)
    result = run(data)
    if result:
        sys.stdout.write(json.dumps(result))
